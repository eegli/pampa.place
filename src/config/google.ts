type GoogleConfig = {
  svRequest: google.maps.StreetViewLocationRequest;
  streetview: google.maps.StreetViewPanoramaOptions;
  map: google.maps.MapOptions;
};

export const config: GoogleConfig = {
  svRequest: {
    radius: 1000,
  },
  streetview: {
    zoom: 0,
    addressControl: false,
    motionTracking: false,
    zoomControl: false,
    showRoadLabels: false,
    fullscreenControl: false,
    pov: {
      heading: 35,
      pitch: 10,
    },
  },
  map: {
    draggableCursor: 'crosshair',
    clickableIcons: false,
    disableDefaultUI: true,
    mapTypeId: 'hybrid',
    mapTypeControl: true,
    gestureHandling: 'auto',
    mapTypeControlOptions: {
      mapTypeIds: ['hybrid', 'roadmap'],
    },
  },
};
