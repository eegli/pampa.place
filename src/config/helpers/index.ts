import Tarea from '@turf/area';
import Tbbox from '@turf/bbox';
import TbboxPolygon from '@turf/bbox-polygon';
import { FeatureCollection, MultiPolygon, Polygon } from '@turf/helpers';
import { nanoid } from 'nanoid';
import {
  MapDataCollection,
  MapIdCollection,
  MapProperties,
  SwissMapProperties,
} from '../types';

export function computeMapData<
  T extends FeatureCollection<
    Polygon | MultiPolygon,
    MapProperties | SwissMapProperties
  >
>(m: T): MapDataCollection {
  return m.features.reduce((acc, curr) => {
    const bb = Tbbox(curr);
    const bbPoly = TbboxPolygon(bb);

    const properties: MapProperties = {
      name: 'unknown',
      category: 'unknown',
    };

    // Narrow down type - Swiss or custom maps
    if ('NAME_LATN' in curr.properties) {
      properties.name = curr.properties.NAME_LATN;
      properties.category = 'switzerland';
    } else {
      properties.name = curr.properties.name;
      properties.category = 'custom';
    }

    delete curr.id;

    // Use an ID in order to avoid collisions between custom and Swiss maps
    acc[nanoid(12)] = {
      feature: {
        ...curr,
        properties,
      },

      area: Tarea(curr.geometry) * 1e-6,
      bb,
      bbLiteral: {
        SW: {
          lng: bbPoly.geometry.coordinates[0][0][0],
          lat: bbPoly.geometry.coordinates[0][0][1],
        },
        SE: {
          lng: bbPoly.geometry.coordinates[0][1][0],
          lat: bbPoly.geometry.coordinates[0][1][1],
        },
        NE: {
          lng: bbPoly.geometry.coordinates[0][2][0],
          lat: bbPoly.geometry.coordinates[0][2][1],
        },
        NW: {
          lng: bbPoly.geometry.coordinates[0][3][0],
          lat: bbPoly.geometry.coordinates[0][3][1],
        },
      },
    };
    acc;
    return acc;
  }, {} as MapDataCollection);
}

export function computeMapIds(m: MapDataCollection) {
  const ids = Object.entries(m).reduce((acc, [mapId, data]) => {
    acc.push({
      name: data.feature.properties.name,
      category: data.feature.properties.category,
      id: mapId,
    });

    return acc;
  }, [] as MapIdCollection);

  return ids.sort((a, b) => (a.name > b.name ? 1 : a.name < b.name ? -1 : 0));
}
