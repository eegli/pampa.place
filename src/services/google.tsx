import {unsafeToggleHTMLElement} from '@/utils/misc';

export const GstvContainer = () => {
  return (
    <div id="__GSTV__CONTAINER__" data-testid="__GSTV__CONTAINER__">
      <div id="__GSTV__" data-testid="__GSTV__" style={{height: '100%'}} />
    </div>
  );
};

export const GmapContainer = () => {
  return (
    <div id="__GMAP__CONTAINER__" data-testid="__GMAP__CONTAINER__">
      <div id="__GMAP__" data-testid="__GMAP__" style={{height: '100%'}} />
    </div>
  );
};

export class StreetViewService {
  static get div() {
    return document.getElementById('__GSTV__')!;
  }
  static get container() {
    return document.getElementById('__GSTV__CONTAINER__')!;
  }
  private static _sv: google.maps.StreetViewPanorama | undefined;

  static get sv() {
    if (!StreetViewService._sv) {
      StreetViewService._sv = new google.maps.StreetViewPanorama(
        StreetViewService.div
      );
      console.log('%cCreated new global Street View instance', 'color: green');
    }
    return StreetViewService._sv;
  }

  static mount<T extends HTMLElement>(tempContainer: T) {
    const unmount = unsafeToggleHTMLElement(
      StreetViewService.div,
      StreetViewService.container,
      tempContainer
    );
    return () => {
      unmount();
    };
  }
}

export class MapService {
  static _map: google.maps.Map | undefined;
  static markers: google.maps.Marker[] = [];

  static addMarker(m: google.maps.Marker) {
    MapService.markers.push(m);
    return m;
  }
  // https://developers.google.com/maps/documentation/javascript/markers#remove
  static clearMarkers() {
    MapService.markers.forEach(m => m.setMap(null));
    MapService.markers.length = 0;
  }

  static get div() {
    return document.getElementById('__GMAP__')!;
  }
  static get container() {
    return document.getElementById('__GMAP__CONTAINER__')!;
  }

  static get map() {
    if (!MapService._map) {
      MapService._map = new google.maps.Map(MapService.div);
      console.log('%cCreated new global Map instance', 'color: green');
    }
    return MapService._map;
  }

  static mount<T extends HTMLElement>(tempContainer: T) {
    const unmount = unsafeToggleHTMLElement(
      MapService.div,
      MapService.container,
      tempContainer
    );

    return () => {
      unmount();
    };
  }
}
