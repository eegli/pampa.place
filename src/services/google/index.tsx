export {GmapContainer, GstvContainer} from './dom';
export {MapService, StreetViewService} from './static';
