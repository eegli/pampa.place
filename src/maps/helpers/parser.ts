import {FeatureParser, UserGeoJSONParser} from '@/config/types';
import Tarea from '@turf/area';
import Tbbox from '@turf/bbox';
import TbboxPolygon from '@turf/bbox-polygon';

export const parseGeoJSONFeature: FeatureParser = (feature, category) => {
  try {
    // Only GeoJSON features "with an area" are allowed in this
    // game - No Points, LineStrings, etc.

    if (!feature.properties?.name) {
      throw new Error(
        `One of the Features (category "${category}") does not have a "name" property.`
      );
    }

    if (!feature.geometry.type.match(/(MultiPolygon|Polygon)$/gi)) {
      throw new Error(
        `Feature "${feature.properties.name}" (category "${category}") is not a Polygon or MultiPolygon.`
      );
    }

    const bb = Tbbox(feature);
    const bbPoly = TbboxPolygon(bb);

    // Maps with the same name AND category will overwrite each other!
    const mapId = `${category}-${feature.properties.name}`
      .replace(/\s/g, '-')
      .toLowerCase();

    return {
      type: feature.type,
      geometry: feature.geometry,
      properties: {
        id: mapId,
        name: feature.properties.name,
        category: category,
        area: Tarea(feature.geometry) * 1e-6,
        bb,
        bbLiteral: {
          SW: {
            lng: bbPoly.geometry.coordinates[0][0][0],
            lat: bbPoly.geometry.coordinates[0][0][1],
          },
          SE: {
            lng: bbPoly.geometry.coordinates[0][1][0],
            lat: bbPoly.geometry.coordinates[0][1][1],
          },
          NE: {
            lng: bbPoly.geometry.coordinates[0][2][0],
            lat: bbPoly.geometry.coordinates[0][2][1],
          },
          NW: {
            lng: bbPoly.geometry.coordinates[0][3][0],
            lat: bbPoly.geometry.coordinates[0][3][1],
          },
        },
      },
    };
  } catch (e) {
    throw new Error(`Failed to compute map data.\n${e}`);
  }
};

// Parse maps that users input in /my-maps. This throws if the input
// is invalid, so it should be wrapped in try/catch whenever user map
// input is parsed. For user input maps, only the first feature in a
// collection is considered!
export const parseUserGeoJSON: UserGeoJSONParser = (raw, name, category) => {
  let parsed = JSON.parse(raw);
  if (Array.isArray(parsed.features)) {
    parsed = parsed.features[0];
  }
  parsed.properties = {name};
  return parseGeoJSONFeature(parsed, category);
};
