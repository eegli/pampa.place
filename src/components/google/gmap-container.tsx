import {MapService} from '@/services/google';
import Box from '@mui/material/Box';
import {ReactNode, useEffect, useRef} from 'react';

type InstanceMethods = Omit<
  InstanceType<typeof google.maps.Map>,
  'setCenter' | 'fitBounds'
>;

export type GoogleMapContainerProps = {
  children: ReactNode;
  onMount?: (map: InstanceMethods) => unknown;
  onUnmount?: (map: InstanceMethods) => unknown;
} & (
  | {
      bounds: google.maps.LatLngLiteral;
      padding?: number;
    }
  | {
      center: google.maps.LatLngLiteral;
      zoom?: number;
    }
);

export const GoogleMapContainer = ({
  children,
  onMount,
  onUnmount,
  ...rest
}: GoogleMapContainerProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmount = MapService.mount(ref.current);

      if ('bounds' in rest) {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(rest.bounds);
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, rest.padding || 0);
        });
      } else {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 0);
      }

      if (typeof onMount === 'function') {
        onMount(MapService.map);
      }

      return () => {
        if (typeof onUnmount === 'function') {
          onUnmount(MapService.map);
        }
        unmount();
      };
    }
  }, [ref, onMount, onUnmount, rest]);

  return (
    <>
      <Box data-testid="play-google-map" ref={ref} height="100%" width="100%" />
      {children}
    </>
  );
};
