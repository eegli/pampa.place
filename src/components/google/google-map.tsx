import {MapService} from '@/services/google';
import Box from '@mui/material/Box';
import {memo, ReactNode, useEffect, useRef} from 'react';

type BoundsProps = {
  bounds: {
    SW: google.maps.LatLngLiteral;
    NE: google.maps.LatLngLiteral;
  };
  padding?: number;
};

type CenterProps = {
  center: google.maps.LatLngLiteral;
  zoom?: number;
};

export type GoogleMapProps = {
  id?: string;
  children?: ReactNode;
  onMount?: (map: google.maps.Map) => unknown;
  onUnmount?: (map: google.maps.Map) => unknown;
} & (BoundsProps | CenterProps);

const _GoogleMap = ({
  children,
  id = 'google-map',
  onMount,
  onUnmount,
  ...rest
}: GoogleMapProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmountMap = MapService.mount(ref.current);
      console.info('%cGOOGLE MAP MOUNT', 'color: yellow');

      if ('bounds' in rest) {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(
          rest.bounds.SW,
          rest.bounds.NE
        );
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, rest.padding || 0);
        });
      } else {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 0);
      }

      if (typeof onMount === 'function') {
        onMount(MapService.map);
      }

      return () => {
        // Any cleanup logic
        if (typeof onUnmount === 'function') {
          onUnmount(MapService.map);
        }

        unmountMap();
        console.info('%cGOOGLE MAP UNMOUNT', 'color: yellow');
      };
    }
  }, [ref, onMount, onUnmount, rest]);

  return (
    <Box id={id} data-testid={id} ref={ref} height="100%" width="100%">
      {children}
    </Box>
  );
};

export const GoogleMap = memo(_GoogleMap, (prev, next) => prev.id === next.id);
