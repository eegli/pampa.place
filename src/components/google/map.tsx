import {svgMarker, svgMarkerColors} from '@/components/google/marker';
import {config} from '@/config/google';
import {LatLngLiteral, MapData} from '@/config/types';
import {Result} from '@/redux/game';
import {useAppDispatch} from '@/redux/hooks';
import {updateSelectedPosition} from '@/redux/position';
import {MapService, MarkerService, PolyLineService} from '@/services/google';
import {useEffect, useRef} from 'react';

export type GoogleMapProps = {
  map: MapData;
  mode?: 'preview' | 'play' | 'result';
  results?: Pick<Result, 'selected' | 'name'>[];
  initialPosition?: LatLngLiteral;
};

export const GoogleMap = ({
  map,
  mode,
  results,
  initialPosition,
}: GoogleMapProps) => {
  const dispatch = useAppDispatch();
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmount = MapService.mount(ref.current);
      /* Order in constructor is important! SW, NE  */
      const bounds = new google.maps.LatLngBounds(
        new google.maps.LatLng(map.properties.bbLiteral.SW),
        new google.maps.LatLng(map.properties.bbLiteral.NE)
      );
      google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
        MapService.map.fitBounds(bounds, 0);
      });

      return () => {
        unmount();
      };
    }
  }, [map]);

  useEffect(() => {
    if (mode === 'preview') {
      console.info('PREVIEW MODE MOUNT');
      MapService.map.setOptions({
        ...config.map,
        gestureHandling: 'none',
        mapTypeId: 'roadmap',
        mapTypeControl: false,
      });

      const features = MapService.map.data.addGeoJson(map);

      MapService.map.data.setStyle({
        fillColor: '#003d80',
        fillOpacity: 0.2,
        strokeWeight: 0.8,
      });
      return () => {
        console.info('PREVIEW MODE UNMOUNT');
        features.forEach(feat => {
          MapService.map.data.remove(feat);
        });
      };
    }
  }, [mode, map]);

  useEffect(() => {
    if (mode === 'play') {
      console.info('PLAY MODE MOUNT');
      MapService.map.setOptions(config.map);

      const marker = new google.maps.Marker();
      marker.setMap(MapService.map);
      marker.setDraggable(true);

      MarkerService.add(marker);

      // This Google Map event is not typed unfortunately
      const listener = MapService.map.addListener('click', (e: unknown) => {
        const {latLng} = e as {latLng: google.maps.LatLng};
        marker.setPosition(latLng);
        dispatch(
          updateSelectedPosition({lat: latLng.lat(), lng: latLng.lng()})
        );
      });
      return () => {
        console.info('PLAY MODE UNMOUNT');
        if (listener) listener.remove();
        // https://developers.google.com/maps/documentation/javascript/markers#remove
        MarkerService.clearAllItems();
      };
    }
  }, [mode, dispatch]);

  useEffect(() => {
    if (mode === 'result' && initialPosition) {
      console.info('RESULT MODE MOUNT');
      MapService.map.setOptions(config.map);

      // Add original location marker
      const originMarker = new google.maps.Marker();
      originMarker.setPosition(initialPosition);
      originMarker.setMap(MapService.map);

      MarkerService.add(originMarker);

      // Add marker for each result
      results &&
        results.forEach((p, idx) => {
          if (p.selected) {
            const playerMarker = new google.maps.Marker({
              position: p.selected,
              map: MapService.map,
              label: {
                text: p.name,
                color: 'white',
                className: 'map-marker',
              },
              icon: {
                path: svgMarker.path,
                fillColor: `#${svgMarkerColors[idx]}`,
                fillOpacity: 1,
                anchor: new google.maps.Point(
                  svgMarker.anchor[0],
                  svgMarker.anchor[1]
                ),
                strokeWeight: 0,
                scale: 1,
                labelOrigin: new google.maps.Point(15, 60),
              },
            });

            const polyLine = new google.maps.Polyline({
              path: [initialPosition, p.selected],
              map: MapService.map,
              geodesic: true,
              strokeColor: `#${svgMarkerColors[idx]}`,
              strokeOpacity: 1.0,
              strokeWeight: 4,
            });
            PolyLineService.add(polyLine);
            MarkerService.add(playerMarker);
          }
        });

      return () => {
        console.info('RESULT MODE UNMOUNT');
        MarkerService.clearAllItems();
        PolyLineService.clearAllItems();
      };
    }
  }, [mode, initialPosition, results]);

  return (
    <div
      ref={ref}
      style={{
        height: '100%',
      }}
    />
  );
};
