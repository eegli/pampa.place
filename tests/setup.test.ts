/* 
There are a lot of Google implementations that are used in the thunks
that need to be mocked properly, e.g. StreetViewPreference and
StreetViewSource (see jest-setup.ts). Make sure they have a correctly
defined initial implementation that can later be overwritten
individually for each test. 
*/

import {parseGeoJSONFeature} from '@/maps/helpers/parser';
import {testMap} from './fixtures/map';
import {GoogleStreetViewResponse} from './payloads/google';

describe('Mock map for tests', () => {
  it('fixtures match computed maps', () => {
    const computed = parseGeoJSONFeature(testMap, 'test');
    expect(computed).toStrictEqual(testMap);
  });
});

describe('Global mock setup', () => {
  it('street view service resolves', async () => {
    const service = new google.maps.StreetViewService();
    expect(service).toBeTruthy();
    expect(await service.getPanorama({})).toStrictEqual(
      GoogleStreetViewResponse
    );
    google.maps.StreetViewService.prototype.getPanorama = jest
      .fn()
      .mockResolvedValue({
        ok: true,
      });

    expect(await service.getPanorama({})).toMatchInlineSnapshot(`
      {
        "ok": true,
      }
    `);
  });
});
