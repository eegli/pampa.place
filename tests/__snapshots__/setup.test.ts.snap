// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test global setup has global definitions 1`] = `
Object {
  "Circle": [Function],
  "ControlPosition": Object {
    "0": "BOTTOM_CENTER",
    "1": "BOTTOM_LEFT",
    "10": "TOP_LEFT",
    "11": "TOP_RIGHT",
    "2": "BOTTOM_RIGHT",
    "3": "LEFT_BOTTOM",
    "4": "LEFT_CENTER",
    "5": "LEFT_TOP",
    "6": "RIGHT_BOTTOM",
    "7": "RIGHT_CENTER",
    "8": "RIGHT_TOP",
    "9": "TOP_CENTER",
    "BOTTOM_CENTER": 0,
    "BOTTOM_LEFT": 1,
    "BOTTOM_RIGHT": 2,
    "LEFT_BOTTOM": 3,
    "LEFT_CENTER": 4,
    "LEFT_TOP": 5,
    "RIGHT_BOTTOM": 6,
    "RIGHT_CENTER": 7,
    "RIGHT_TOP": 8,
    "TOP_CENTER": 9,
    "TOP_LEFT": 10,
    "TOP_RIGHT": 11,
  },
  "Data": [Function],
  "ImageMapType": [MockFunction],
  "LatLng": [Function],
  "LatLngBounds": [Function],
  "MVCArray": [Function],
  "MVCObject": [Function],
  "Map": [Function],
  "MapCanvasProjection": [Function],
  "MapPanes": [Function],
  "MapTypeId": Object {
    "HYBRID": "hybrid",
    "ROADMAP": "roadmap",
    "SATELLITE": "satellite",
    "TERRAIN": "terrain",
  },
  "Marker": [Function],
  "OverlayView": [Function],
  "Point": [Function],
  "Polyline": [Function],
  "Size": [Function],
  "StreetViewCoverageLayer": [Function],
  "StreetViewPanorama": [Function],
  "StreetViewPreference": Object {
    "BEST": "best",
    "NEAREST": "nearest",
  },
  "StreetViewService": [Function],
  "StreetViewSource": Object {
    "DEFAULT": "default",
    "OUTDOOR": "outdoor",
  },
  "StreetViewStatus": Object {
    "OK": "OK",
    "UNKNOWN_ERROR": "UNKNOWN_ERROR",
    "ZERO_RESULTS": "ZERO_RESULTS",
  },
  "VisibleRegion": [Function],
  "event": Object {
    "addDomListener": [MockFunction],
    "addDomListenerOnce": [MockFunction],
    "addListener": [MockFunction],
    "addListenerOnce": [MockFunction],
    "clearInstanceListeners": [MockFunction],
    "clearListeners": [MockFunction],
    "hasListeners": [MockFunction],
    "removeListener": [MockFunction],
    "trigger": [MockFunction],
  },
  "places": Object {
    "Autocomplete": [Function],
    "SearchBox": [Function],
  },
}
`;
